popgen/mcmcSelection.go

```diff
@@ -122,20 +127,35 @@ func MetropolisHastings(data AFS, muZero float64, sigmaZero float64, iterations
 	}
 
 	maxN := findMaxN(data)
-	binomMap := make([][]float64, maxN+1)
+	allN := findAllN(data)
+	var alleleFrequencyCache []float64 = cacheIndexToAlleleFrequencyCache(bins)
+	nkpCache := make([][][]float64, maxN+1)
+	var coefficient float64
+
+	for n := 0; n < len(allN); n++ {
+		nkpCache[allN[n]] = make([][]float64, allN[n])
+		for k := 1; k < allN[n]; k++ {
```

Comment:
This triple nested loop gets a little crazy allocating memory for k and p. Although it is possible the complier performs these optimizations during run time, it's very possible it doesn't and if that is the case, you are basically do an extra k*n operations by doing` k := 1` and `p := 0` under the n loop.

Reply:
I'm not sure I understand where there is an extra calculation, could you expand a bit on this?

Reply:
because you call k := 1 under the `for n := 0; n < len(allN); n++ {` you allocate new memory for every new iteration of n. whereas simply doing k = 0 (no colon) you reuse the same memory. it's very subtle but k := 1 and k = 1 behaves a little different

Reply:
on top of this there is the p := 0, under the k for loop. For for every iteration of K, calling p := 0 is different from p = 0. `p := 0` is called k times which results in another n memory allocations.
in summary doing something like:
```
var n, k, p int
for n = 0; n < len; n++ {
	for k = 0; k < len; k ++ {
		for p = 0; p < len; p++ {
				
		}
	}
} 
```
is more desirable than:
```
for n  := 0; n < len; n++ {
	for k  := 0; k < len; k ++ {
		for p := 0; p < len; p++ {
				
		}
	}
}

Reply:
if you don't believe me, here is a quick benchmark function you can
`go test -bench=.` gives us:
BenchmarkTripleLoopWithColon-8   	`1997708,	       586 ns/op`
BenchmarkTripleLoopWithout-8       `3108403,	       383 ns/op`
gives us roughly a 30% boost in performance
```
package main

import (
	"testing"
)

func BenchmarkTripleLoopWithColon(b *testing.B) {
	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		WithColon(10,10,10)
	}
}

func BenchmarkTripleLoopWithout(b *testing.B) {
	b.ResetTimer()
	for n := 0; n < b.N; n++ {
		WithOut(10, 10, 10)
	}
}

func WithColon(one, two, three int) int{
	var ans int = 0
	for i := 0; i < one; i++ {
		for j := 0; j < two; j++ {
			for k := 0; k < three; k++ {
				ans+=1
			}
		}
	}
	return ans
}

func WithOut(one, two, three int) int{
	var ans int = 0
	var i, j, k int
	for i = 0; i < one; i++ {
		for j = 0; j < two; j++ {
			for k = 0; k < three; k++ {
				ans+=1
			}
		}
	}
	return ans
}
```

Reply:
I think there is a good general point here that the ":=" operator is quick and easy when writing code, but can at times have drawbacks when performance is important.  One thing that might be good is to go through and look for ":=" within loops and replace it by declaring the variable once at the top of the function.

Reply:
Great, thanks!